File: apple/src/FFmpegKitInputBuffer.m
/*
 * Copyright (c) 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

#import "FFmpegKitInputBuffer.h"
#import "FFmpegKitConfig.h"

@interface FFmpegKitConfig (FFmpegKitBufferSupport)
+ (long)registerFFmpegKitInputBuffer:(NSData *)data;
+ (long)registerFFmpegKitInputBufferWithBytes:(const void *)bytes
                                       length:(NSUInteger)length;
+ (long)getFFmpegKitBufferSize:(long)bufferId;
+ (void)unregisterFFmpegKitBuffer:(long)bufferId;
@end

@interface FFmpegKitInputBuffer ()
+ (NSString *)urlWithProtocol:(NSString *)protocol
                   resourceId:(long)resourceId
                    extension:(NSString *)extension;
- (instancetype)initWithResourceId:(long)resourceId
                         extension:(NSString *)extension;
- (void)ensureOpen;
@end

@implementation FFmpegKitInputBuffer {
    long _resourceId;
    NSString *_url;
    BOOL _closed;
}

+ (instancetype)fromData:(NSData *)data extension:(NSString *)extension {
    if (data == nil) {
        [NSException raise:NSInvalidArgumentException
                    format:@"data must not be nil"];
    }

    long resourceId = [FFmpegKitConfig registerFFmpegKitInputBuffer:data];
    if (resourceId == 0) {
        [NSException raise:NSInternalInconsistencyException
                    format:@"Failed to register FFmpegKit input buffer."];
    }

    return [[self alloc] initWithResourceId:resourceId extension:extension];
}

+ (instancetype)fromBytes:(const void *)bytes
                   length:(NSUInteger)length
                extension:(NSString *)extension {
    if (bytes == NULL && length > 0) {
        [NSException raise:NSInvalidArgumentException
                    format:@"bytes must not be NULL when length is positive"];
    }

    long resourceId =
        [FFmpegKitConfig registerFFmpegKitInputBufferWithBytes:bytes
                                                        length:length];
    if (resourceId == 0) {
        [NSException raise:NSInternalInconsistencyException
                    format:@"Failed to register FFmpegKit input buffer."];
    }

    return [[self alloc] initWithResourceId:resourceId extension:extension];
}

+ (NSString *)urlWithProtocol:(NSString *)protocol
                   resourceId:(long)resourceId
                    extension:(NSString *)extension {
    NSString *normalizedExtension = @"bin";

    if (extension != nil) {
        NSString *candidate = [[extension
            stringByTrimmingCharactersInSet:
                [NSCharacterSet whitespaceAndNewlineCharacterSet]]
            lowercaseStringWithLocale:[NSLocale localeWithLocaleIdentifier:
                                                   @"en_US_POSIX"]];

        while ([candidate hasPrefix:@"."]) {
            candidate = [candidate substringFromIndex:1];
        }

        NSMutableString *safeExtension = [[NSMutableString alloc] init];
        for (NSUInteger i = 0; i < [candidate length]; i++) {
            unichar c = [candidate characterAtIndex:i];
            if ((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9')) {
                [safeExtension appendFormat:@"%C", c];
            }
        }

        if ([safeExtension length] > 0) {
            normalizedExtension = safeExtension;
        }
    }

    return [NSString stringWithFormat:@"%@:%ld.%@", protocol, resourceId,
                                      normalizedExtension];
}

- (instancetype)initWithResourceId:(long)resourceId
                         extension:(NSString *)extension {
    self = [super init];
    if (self) {
        _resourceId = resourceId;
        _url = [FFmpegKitInputBuffer urlWithProtocol:@"ffkitmem"
                                          resourceId:resourceId
                                           extension:extension];
        _closed = NO;
    }

    return self;
}

- (NSString *)getUrl {
    [self ensureOpen];
    return _url;
}

- (long)getSize {
    [self ensureOpen];
    return [FFmpegKitConfig getFFmpegKitBufferSize:_resourceId];
}

- (void)close {
    if (!_closed) {
        [FFmpegKitConfig unregisterFFmpegKitBuffer:_resourceId];
        _closed = YES;
    }
}

- (void)ensureOpen {
    if (_closed) {
        [NSException raise:NSInternalInconsistencyException
                    format:@"FFmpegKit input buffer is closed."];
    }
}

@end

File: windows/src/internal/FFmpegKitInputBuffer.cpp
/*
 * Copyright (c) 2026 Taner Sener
 *
 * This file is part of FFmpegKitNext.
 *
 * FFmpegKitNext is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKitNext is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 * You should have received a copy of the GNU Lesser General License
 * along with FFmpegKitNext. If not, see <http://www.gnu.org/licenses/>.
 */

#include "FFmpegKitInputBuffer.h"
#include "FFmpegKitConfig.h"
#include "FFmpegKitProtocolUrl.h"
#include <stdexcept>

std::shared_ptr<ffmpegkit::internal::FFmpegKitInputBuffer>
ffmpegkit::internal::FFmpegKitInputBuffer::fromByteArray(
    const std::vector<uint8_t> &data, const std::string &extension) {
    long id = ffmpegkit::internal::FFmpegKitConfig::registerFFmpegKitInputBuffer(data);
    if (id == 0) {
        throw std::runtime_error("Failed to register FFmpegKit input buffer.");
    }

    return std::shared_ptr<ffmpegkit::internal::FFmpegKitInputBuffer>(
        new ffmpegkit::internal::FFmpegKitInputBuffer(id, extension));
}

std::shared_ptr<ffmpegkit::internal::FFmpegKitInputBuffer>
ffmpegkit::internal::FFmpegKitInputBuffer::fromBytes(const uint8_t *data,
                                           const size_t size,
                                           const std::string &extension) {
    if (data == NULL && size > 0) {
        throw std::invalid_argument(
            "data must not be null when size is positive");
    }

    long id =
        ffmpegkit::internal::FFmpegKitConfig::registerFFmpegKitInputBuffer(data, size);
    if (id == 0) {
        throw std::runtime_error("Failed to register FFmpegKit input buffer.");
    }

    return std::shared_ptr<ffmpegkit::internal::FFmpegKitInputBuffer>(
        new ffmpegkit::internal::FFmpegKitInputBuffer(id, extension));
}

ffmpegkit::internal::FFmpegKitInputBuffer::~FFmpegKitInputBuffer() { close(); }

ffmpegkit::internal::FFmpegKitInputBuffer::FFmpegKitInputBuffer(
    const long id, const std::string &extension)
    : _id{id}, _url{ffmpegkit::internal::buildFFmpegKitUrl("ffkitmem", id, extension)},
      _closed{false} {}

std::string ffmpegkit::internal::FFmpegKitInputBuffer::getUrl() {
    ensureOpen();
    return _url;
}

long ffmpegkit::internal::FFmpegKitInputBuffer::getSize() {
    ensureOpen();
    return ffmpegkit::internal::FFmpegKitConfig::getFFmpegKitBufferSize(_id);
}

void ffmpegkit::internal::FFmpegKitInputBuffer::close() {
    if (!_closed) {
        ffmpegkit::internal::FFmpegKitConfig::unregisterFFmpegKitBuffer(_id);
        _closed = true;
    }
}

void ffmpegkit::internal::FFmpegKitInputBuffer::ensureOpen() {
    if (_closed) {
        throw std::runtime_error("FFmpegKit input buffer is closed.");
    }
}
